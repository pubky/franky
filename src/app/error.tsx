'use client';

import { AppError, CommonErrorType } from '@/lib/error';
import { useEffect } from 'react';

export default function Error({ error, reset }: { error: Error; reset: () => void }) {
  useEffect(() => {
    // Convert to AppError if it's not already one
    if (!(error instanceof AppError)) {
      new AppError(CommonErrorType.UNEXPECTED_ERROR, error.message || 'An unexpected error occurred', 500, {
        originalError: error,
      });
    }
  }, [error]);

  return (
    <div className="flex min-h-screen items-center justify-center bg-background p-5">
      <div className="w-full max-w-2xl rounded-lg bg-card p-8 text-center shadow-md">
        <h1 className="mb-4 text-xl font-semibold text-destructive">Something went wrong</h1>
        <p className="mb-6 text-muted-foreground">
          {error instanceof AppError ? error.message : 'An unexpected error occurred'}
        </p>
        {process.env.NODE_ENV === 'development' && (
          <details className="mb-6 text-left">
            <summary className="cursor-pointer text-primary hover:underline">Error Details</summary>
            <pre className="mt-2 overflow-x-auto rounded-md bg-muted p-4 text-sm">
              {error instanceof AppError ? JSON.stringify(error, null, 2) : error.toString()}
            </pre>
          </details>
        )}
        <button
          onClick={reset}
          className="rounded-md bg-primary px-4 py-2 text-primary-foreground transition-colors hover:bg-primary/90"
        >
          Try Again
        </button>
      </div>
    </div>
  );
}
