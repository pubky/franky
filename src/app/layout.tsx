import { Inter_Tight } from 'next/font/google';
import { DatabaseProvider } from '@/providers';

import './globals.css';

import * as Organisms from '@/organisms';
import * as Molecules from '@/molecules';
import * as Atoms from '@/atoms';

const interTight = Inter_Tight({
  variable: '--font-geist-sans',
  subsets: ['latin'],
});

export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <RootContainer>
      <DatabaseProvider>
        <Molecules.Metadata />
        <Organisms.Header />
        {children}
        <Molecules.Toaster />
      </DatabaseProvider>
    </RootContainer>
  );
}

export function RootContainer({ children }: { children: React.ReactNode }) {
  return (
    <Atoms.Container as="html" lang="en" className="dark">
      <Atoms.Container as="body" className={`${interTight.variable} antialiased`}>
        <Molecules.PageContainer>{children}</Molecules.PageContainer>
      </Atoms.Container>
    </Atoms.Container>
  );
}
