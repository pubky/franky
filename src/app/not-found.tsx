import * as Molecules from '@/molecules';

export const metadata = Molecules.Metadata({
  title: 'Page Not Found - Pubky App',
  description: 'The page you are looking for does not exist.',
  robots: false,
});

export default function NotFound() {
  return (
    <div className="flex h-screen flex-col items-center justify-center">
      <h2 className="text-2xl font-bold">Not Found</h2>
      <p>Could not find requested resource</p>
    </div>
  );
}
