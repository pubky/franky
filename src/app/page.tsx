'use client';

import { Header, HeroSection } from '@/components/ui';

export default function Home() {
  const handleSignIn = () => {
    console.log('Sign in clicked');
  };

  const handleCreateAccount = () => {
    console.log('Create account clicked');
  };

  return (
    <>
      {/* Full viewport background */}
      <div
        className={`opacity-10 lg:opacity-100 fixed inset-0 bg-cover bg-center bg-no-repeat -z-10`}
        style={{
          backgroundImage: 'url(/images/bg-home.svg)',
        }}
      />

      <div className="dark min-h-screen bg-transparent z-10">
        <Header onSignIn={handleSignIn} githubUrl="#" twitterUrl="#" telegramUrl="#" />

        <main className="relative">
          <HeroSection
            title={
              <>
                <span className="text-brand">Unlock</span>
                <br />
                the web.
              </>
            }
            subtitle="Pubky requires an invite code"
            onSignIn={handleSignIn}
            onCreateAccount={handleCreateAccount}
          />
        </main>
      </div>
    </>
  );
}
