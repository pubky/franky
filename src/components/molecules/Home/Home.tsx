'use client';

import { useRouter } from 'next/navigation';

import * as Molecules from '@/molecules';
import * as Atoms from '@/atoms';
import * as Libs from '@/libs';
import { PUBKY_CORE_URL } from '@/config';

export const HomeActions = () => {
  const router = useRouter();

  const handleCreateAccount = () => {
    router.push('/onboarding/install');
  };

  const handleSignIn = () => {
    console.log('Sign in clicked');
  };

  return <Molecules.ActionButtons onSignIn={handleSignIn} onCreateAccount={handleCreateAccount} />;
};

export const HomeFooter = ({ ...props }: React.HTMLAttributes<HTMLDivElement>) => {
  return (
    <Atoms.FooterLinks
      className={Libs.cn('h-26 sm:pr-16 justify-end flex-row content-end', props.className)}
      {...props}
    >
      By creating a Pubky account, you agree to the <Molecules.DialogTerms />, <Molecules.DialogPrivacy />, and confirm
      you are <Molecules.DialogAge /> Pubky is powered by{' '}
      <Atoms.Link href={PUBKY_CORE_URL} target="_blank">
        Pubky Core
      </Atoms.Link>{' '}
      and was built with love and dedication by Synonym Software Ltd. Â©2025.
    </Atoms.FooterLinks>
  );
};

export const HomeSectionTitle = () => {
  return (
    <Atoms.Container className="flex-row items-start gap-2">
      <Atoms.Typography size="md" className="text-brand font-light sm:text-2xl self-center">
        Pubky requires an invite code
      </Atoms.Typography>
      <Molecules.PopoverInvite />
    </Atoms.Container>
  );
};

export const HomePageHeading = () => {
  return (
    <Atoms.Heading level={1} size="2xl">
      <span className="text-brand">Unlock</span> <br className="block sm:hidden" />
      the web.
    </Atoms.Heading>
  );
};
