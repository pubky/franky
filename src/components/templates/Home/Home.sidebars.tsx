import * as Molecules from '@/molecules';
import * as Libs from '@/libs';
import * as Core from '@/core';

export interface HomeSidebarProps {
  reach: Core.ReachType;
  setReach: (reach: Core.ReachType) => void;
  sort: Core.SortType;
  setSort: (sort: Core.SortType) => void;
  content: Core.ContentType;
  setContent: (content: Core.ContentType) => void;
  layout: Core.LayoutType;
  setLayout: (layout: Core.LayoutType) => void;
}

export function HomeLeftSidebar({
  reach,
  setReach,
  sort,
  setSort,
  content,
  setContent,
  layout,
  setLayout,
}: HomeSidebarProps) {
  return (
    <>
      <Molecules.FilterReach selectedTab={reach} onTabChange={setReach} />
      <Molecules.FilterSort selectedTab={sort} onTabChange={setSort} />
      <div className="self-start sticky top-[100px] flex flex-col gap-6">
        <Molecules.FilterContent selectedTab={content} onTabChange={setContent} />
        <Molecules.FilterLayout selectedTab={layout} onTabChange={setLayout} />
      </div>
    </>
  );
}

export function HomeRightSidebar() {
  return (
    <>
      <Molecules.WhoToFollow />
      <Molecules.ActiveUsers />
      <Molecules.HotTags
        tags={[
          { name: 'bitcoin', count: 1234 },
          { name: 'nostr', count: 892 },
          { name: 'decentralization', count: 567 },
          { name: 'privacy', count: 445 },
          { name: 'web3', count: 321 },
          { name: 'opensource', count: 289 },
        ]}
      />
      <div className="self-start sticky top-[100px]">
        <Molecules.FeedbackCard />
      </div>
    </>
  );
}

export function HomeLeftDrawer({
  reach,
  setReach,
  sort,
  setSort,
  content,
  setContent,
  layout,
  setLayout,
}: HomeSidebarProps) {
  return (
    <div className="flex flex-col gap-6">
      <Molecules.FilterReach selectedTab={reach} onTabChange={setReach} />
      <Molecules.FilterSort selectedTab={sort} onTabChange={setSort} />
      <Molecules.FilterContent selectedTab={content} onTabChange={setContent} />
      <Molecules.FilterLayout selectedTab={layout} onTabChange={setLayout} />
    </div>
  );
}

export function HomeRightDrawer() {
  return (
    <div className="flex flex-col gap-6">
      <Molecules.WhoToFollow />
      <Molecules.ActiveUsers />
      <Molecules.HotTags
        tags={[
          { name: 'bitcoin', count: 1234 },
          { name: 'nostr', count: 892 },
          { name: 'decentralization', count: 567 },
          { name: 'privacy', count: 445 },
          { name: 'web3', count: 321 },
          { name: 'opensource', count: 289 },
        ]}
      />
      <Molecules.FeedbackCard />
    </div>
  );
}

export function HomeLeftDrawerMobile({
  reach,
  setReach,
  sort,
  setSort,
  content,
  setContent,
}: Omit<HomeSidebarProps, 'layout' | 'setLayout'>) {
  return (
    <div className="flex flex-col gap-6">
      <Molecules.FilterReach selectedTab={reach} onTabChange={setReach} />
      <Molecules.FilterSort selectedTab={sort} onTabChange={setSort} />
      <Molecules.FilterContent selectedTab={content} onTabChange={setContent} />
    </div>
  );
}

export function HomeRightDrawerMobile() {
  return (
    <Molecules.FeedSection
      feeds={[
        { icon: Libs.UsersRound, label: 'Following' },
        { icon: Libs.Pencil, label: 'Based bitcoin' },
        { icon: Libs.Pencil, label: 'Mining industry' },
      ]}
      showCreateButton={true}
    />
  );
}
