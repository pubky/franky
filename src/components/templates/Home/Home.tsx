import * as Atoms from '@/atoms';
import * as Molecules from '@/molecules';

export function Home() {
  return (
    <>
      <Atoms.ImageBackground image="/images/bg-home.svg" mobileImage="/images/bg-home-mobile.svg" />
      <Atoms.Container size="container" className="px-6">
        <Molecules.PageContainer size="narrow" className="items-start mx-0 flex flex-col gap-6">
          <Molecules.HomePageHeading />
          <Molecules.HomeSectionTitle />
          <Molecules.HomeActions />
          <Molecules.HomeFooter />
        </Molecules.PageContainer>
      </Atoms.Container>
    </>
  );
}
