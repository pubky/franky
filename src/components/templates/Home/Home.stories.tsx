import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import { Home } from './Home';

const meta: Meta<typeof Home> = {
  title: 'Templates/Home',
  component: Home,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component:
          'The main home template that displays posts from users you follow. Includes placeholder content and a logout button.',
      },
    },
  },
  tags: ['autodocs'],
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {},
  parameters: {
    nextjs: {
      appDirectory: true,
    },
  },
};
