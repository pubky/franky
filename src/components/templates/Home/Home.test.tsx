import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { Home } from './Home';
import * as App from '@/app';

// Mock next/navigation - REQUIRED: external dependency
vi.mock('next/navigation', () => ({
  useRouter: vi.fn(() => ({
    push: vi.fn(),
    replace: vi.fn(),
    prefetch: vi.fn(),
    back: vi.fn(),
    forward: vi.fn(),
    refresh: vi.fn(),
  })),
}));

// Mock dexie-react-hooks - REQUIRED: database dependency
vi.mock('dexie-react-hooks', () => ({
  useLiveQuery: vi.fn(() => null),
}));

// Mock the Core module - REQUIRED: stores and controllers
vi.mock('@/core', async (importOriginal) => {
  const actual = await importOriginal<typeof import('@/core')>();
  return {
    ...actual,
    useAuthStore: vi.fn(() => ({
      currentUserPubky: null,
    })),
    useOnboardingStore: vi.fn(() => ({
      secretKey: null,
      showWelcomeDialog: false,
      setShowWelcomeDialog: vi.fn(),
    })),
    useFiltersStore: vi.fn(() => ({
      layout: 'columns',
      setLayout: vi.fn(),
      reach: 'all',
      setReach: vi.fn(),
      sort: 'recent',
      setSort: vi.fn(),
      content: 'all',
      setContent: vi.fn(),
    })),
    PostController: {
      read: vi.fn(() => Promise.resolve([])),
    },
    db: {
      user_details: {
        get: vi.fn(() => Promise.resolve(null)),
      },
    },
  };
});

// Mock organisms - REQUIRED: complex components with their own dependencies
vi.mock('@/organisms', () => ({
  Post: () => <div data-testid="post">Mocked Post</div>,
  ContentLayout: ({ children }: { children: React.ReactNode }) => <div data-testid="content-layout">{children}</div>,
  DialogWelcome: ({ isOpen, onOpenChange }: { isOpen: boolean; onOpenChange: (open: boolean) => void }) =>
    isOpen ? (
      <div data-testid="dialog-welcome" onClick={() => onOpenChange(false)}>
        Mocked DialogWelcome
      </div>
    ) : null,
  AlertBackup: ({ onDismiss }: { onDismiss?: () => void }) => (
    <div data-testid="alert-backup" onClick={onDismiss}>
      Mocked AlertBackup
    </div>
  ),
}));

// Mock hooks - REQUIRED: custom hooks with complex logic
vi.mock('@/hooks', () => ({
  useInfiniteScroll: vi.fn(() => ({
    sentinelRef: { current: null },
  })),
}));

describe('Home', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders without errors', () => {
    render(<Home />);
    expect(screen.getByText('Home')).toBeInTheDocument();
  });

  it('displays the Home heading correctly', () => {
    render(<Home />);
    const heading = screen.getByText('Home');
    expect(heading).toBeInTheDocument();
  });

  it('displays welcome message', () => {
    render(<Home />);
    expect(
      screen.getByText("Welcome to your home. This is where you'll see posts from people you follow."),
    ).toBeInTheDocument();
  });

  it('renders logout button', () => {
    render(<Home />);
    const logoutButton = screen.getByText('Logout');
    expect(logoutButton).toBeInTheDocument();
    expect(logoutButton).toHaveAttribute('id', 'home-logout-btn');
  });

  it('navigates to logout page when logout button is clicked', async () => {
    const { useRouter } = await import('next/navigation');
    const mockPush = vi.fn();
    (useRouter as ReturnType<typeof vi.fn>).mockReturnValue({ push: mockPush });

    render(<Home />);
    const logoutButton = screen.getByText('Logout');
    fireEvent.click(logoutButton);

    expect(mockPush).toHaveBeenCalledWith(App.AUTH_ROUTES.LOGOUT);
  });

  it('renders container structure correctly', () => {
    render(<Home />);
    expect(screen.getByTestId('content-layout')).toBeInTheDocument();
  });

  describe('Welcome Dialog Behavior', () => {
    it('should not show welcome dialog when showWelcomeDialog is false', async () => {
      const mockUseOnboardingStore = vi.fn(() => ({
        showWelcomeDialog: false,
        setShowWelcomeDialog: vi.fn(),
      }));

      // Update the mock for this test
      const Core = await import('@/core');
      vi.mocked(Core.useOnboardingStore).mockImplementation(mockUseOnboardingStore);

      render(<Home />);

      expect(screen.queryByTestId('dialog-welcome')).not.toBeInTheDocument();
    });

    it('should show welcome dialog when showWelcomeDialog is true and user details exist', async () => {
      const mockSetShowWelcomeDialog = vi.fn();
      const mockUseOnboardingStore = vi.fn(() => ({
        showWelcomeDialog: true,
        setShowWelcomeDialog: mockSetShowWelcomeDialog,
      }));

      const mockUseAuthStore = vi.fn(() => ({
        currentUserPubky: 'test-pubky-123',
      }));

      // Mock useLiveQuery to return user details
      const { useLiveQuery } = await import('dexie-react-hooks');
      vi.mocked(useLiveQuery).mockReturnValue({
        name: 'Test User',
        bio: 'Test bio',
        image: 'test-image.jpg',
      });

      // Update the mocks for this test
      const Core = await import('@/core');
      vi.mocked(Core.useOnboardingStore).mockImplementation(mockUseOnboardingStore);
      vi.mocked(Core.useAuthStore).mockImplementation(mockUseAuthStore);

      render(<Home />);

      expect(screen.getByTestId('dialog-welcome')).toBeInTheDocument();
    });

    it('should not show welcome dialog when user details do not exist', async () => {
      const mockUseOnboardingStore = vi.fn(() => ({
        showWelcomeDialog: true,
        setShowWelcomeDialog: vi.fn(),
      }));

      const mockUseAuthStore = vi.fn(() => ({
        currentUserPubky: null,
      }));

      // Update the mocks for this test
      const Core = await import('@/core');
      vi.mocked(Core.useOnboardingStore).mockImplementation(mockUseOnboardingStore);
      vi.mocked(Core.useAuthStore).mockImplementation(mockUseAuthStore);

      render(<Home />);

      expect(screen.queryByTestId('dialog-welcome')).not.toBeInTheDocument();
    });

    it('should call setShowWelcomeDialog(false) when welcome dialog is closed', async () => {
      const mockSetShowWelcomeDialog = vi.fn();
      const mockUseOnboardingStore = vi.fn(() => ({
        showWelcomeDialog: true,
        setShowWelcomeDialog: mockSetShowWelcomeDialog,
      }));

      const mockUseAuthStore = vi.fn(() => ({
        currentUserPubky: 'test-pubky-123',
      }));

      // Mock useLiveQuery to return user details
      const { useLiveQuery } = await import('dexie-react-hooks');
      vi.mocked(useLiveQuery).mockReturnValue({
        name: 'Test User',
        bio: 'Test bio',
        image: 'test-image.jpg',
      });

      // Update the mocks for this test
      const Core = await import('@/core');
      vi.mocked(Core.useOnboardingStore).mockImplementation(mockUseOnboardingStore);
      vi.mocked(Core.useAuthStore).mockImplementation(mockUseAuthStore);

      render(<Home />);

      const welcomeDialog = screen.getByTestId('dialog-welcome');
      expect(welcomeDialog).toBeInTheDocument();

      // Click to close the dialog
      fireEvent.click(welcomeDialog);

      expect(mockSetShowWelcomeDialog).toHaveBeenCalledWith(false);
    });

    it('should have correct welcome dialog props when shown', async () => {
      const mockSetShowWelcomeDialog = vi.fn();
      const mockUseOnboardingStore = vi.fn(() => ({
        showWelcomeDialog: true,
        setShowWelcomeDialog: mockSetShowWelcomeDialog,
      }));

      const mockUseAuthStore = vi.fn(() => ({
        currentUserPubky: 'test-pubky-123',
      }));

      const mockUserDetails = {
        name: 'Test User',
        bio: 'Test bio',
        image: 'test-image.jpg',
      };

      // Mock useLiveQuery to return user details
      const { useLiveQuery } = await import('dexie-react-hooks');
      vi.mocked(useLiveQuery).mockReturnValue(mockUserDetails);

      // Update the mocks for this test
      const Core = await import('@/core');
      vi.mocked(Core.useOnboardingStore).mockImplementation(mockUseOnboardingStore);
      vi.mocked(Core.useAuthStore).mockImplementation(mockUseAuthStore);

      render(<Home />);

      // The dialog should be rendered with correct props
      // This is implicitly tested by the dialog being shown when conditions are met
      expect(screen.getByTestId('dialog-welcome')).toBeInTheDocument();
    });
  });
});
