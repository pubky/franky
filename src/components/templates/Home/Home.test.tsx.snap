// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home - Snapshots > matches snapshot for default Home 1`] = `
<div>
  <div
    class="opacity-10 lg:opacity-100"
    data-image="/images/bg-home.svg"
    data-testid="image-background"
  >
    Image Background
  </div>
  <div
    class="container container"
    data-testid="container"
  >
    <div
      data-testid="page-container"
    >
      <div
        data-testid="home-page-heading"
      />
      <div
        data-testid="home-section-title"
      >
        Section Title
      </div>
      <div
        data-testid="home-actions"
      >
        Actions
      </div>
      <div
        data-testid="home-footer"
      >
        Footer
      </div>
    </div>
  </div>
</div>
`;
